poly/sharing_test.go

```diff
@@ -0,0 +1,544 @@
+
+	testPubPoly := producePubPoly(g, k, n, s, p)
+	return new(PubShares).Split(testPubPoly, t)
+}
+
+/* Test Functions */
+
+// Test that the Pick function creates unique polynomials and unique secrets.
+func TestPriPolyPick_UniqueShares(t *testing.T) {
+	testPoly1, testPoly2, testPoly3 := produceThreePriPoly(group, k, nil)
+	if testPoly1.Equal(testPoly2) || testPoly1.Equal(testPoly3) ||
+		testPoly2.Equal(testPoly3) {
+		t.Error("Failed to create unique polynomials.")
+	}
+	if testPoly1.Secret().Equal(testPoly2.Secret()) ||
+		testPoly1.Secret().Equal(testPoly3.Secret()) ||
+		testPoly2.Secret().Equal(testPoly3.Secret()) {
+		t.Error("Failed to create unique secrets.")
+	}
+}
+
+// Test polynomials that are based on common secrets. Verify that
+// unique polynomials are created but that the base secrets are all the same.
+func TestPriPolyPick_CommonShares(t *testing.T) {
+	testPoly1, testPoly2, testPoly3 := produceThreePriPoly(group, k, secret)
+	if testPoly1.Equal(testPoly2) || testPoly1.Equal(testPoly3) ||
+		testPoly2.Equal(testPoly3) {
+		t.Error("Failed to create unique polynomials.")
+	}
+	if !testPoly1.Secret().Equal(testPoly2.Secret()) ||
+		!testPoly1.Secret().Equal(testPoly3.Secret()) ||
+		!testPoly2.Secret().Equal(testPoly3.Secret()) {
+		t.Error("Polynomials are expected to have the same secret.")
+	}
+}
+
+// Verify that the secret function works. The function should return the same
+// secret it was initialized with.
+func TestPriPolySecret(t *testing.T) {
+	testPoly := producePriPoly(group, k, secret)
+	if !secret.Equal(testPoly.Secret()) {
+		t.Error("This secret differs from the one given to it.")
+	}
+}
+
+// Verify that Equal returns true for two polynomials that are the same
+func TestPriPolyEqual_Same(t *testing.T) {
+	testPoly := producePriPoly(group, k, secret)
+	testPolyCopy := testPoly
+	if !testPoly.Equal(testPolyCopy) {
+		t.Error("Polynomials are expected to be equal.")
+	}
+}
+
+// Verify that Equal returns false for two polynomials that are diffferent
+func TestPriPolyEqual_Different(t *testing.T) {
+	testPoly1, testPoly2 := produceTwoPriPoly(group, k, secret)
+	if testPoly1.Equal(testPoly2) {
+		t.Error("Polynomials are expected to be different.")
+	}
+}
+
+// Verify that Equal panics if the polynomials are of different degrees.
+func TestPriPolyEqual_Error1(t *testing.T) {
+	defer deferTest(t, "The Equal method should have panicked.")
+	testPoly1 := producePriPoly(group, k, secret)
+	testPoly2 := producePriPoly(group, k+10, secret)
+	testPoly1.Equal(testPoly2)
+}
+
+// Verify that Equal panics if the polynomials are of different groups.
+func TestPriPolyEqual_Error2(t *testing.T) {
+	defer deferTest(t, "The Equal method should have panicked.")
+	testPoly1 := producePriPoly(group, k, secret)
+	testPoly2 := producePriPoly(altGroup, k, altSecret)
+	testPoly1.Equal(testPoly2)
+}
+
+// Verify that the add function properly adds two polynomials together.
+func TestPriPolyAdd_Success(t *testing.T) {
```

Comment:
Combine the _Success, _Error1, _Error2 variants here and in other functions similarly to above.



Also, instead of re-creating a bunch of standard "test polynomials" in each of a bunch of test functions, you can just move those out to global variables in this _test.go module.  Those variables will only ever exist in versions of this package compiled for testing, so they won't pollute the namespace or create any binary code-bloat when this package is used "for real" by other packages.